symbolic-base/src/ZkFold/Symbolic/Data/Class.hs

```diff
@@ -4,287 +4,160 @@
-import           Data.Functor                ((<$>))
-import           Data.Functor.Rep            (Representable)
-import qualified Data.Functor.Rep            as R
-import           Data.Kind                   (Type)
-import           Data.Traversable            (Traversable)
-import           Data.Tuple                  (fst)
-import           Data.Type.Equality          (type (~))
-import           Data.Typeable               (Proxy (..))
-import           GHC.Generics                (U1 (..), (:*:) (..), (:.:) (..))
-import qualified GHC.Generics                as G
-
-import           ZkFold.Algebra.Number       (KnownNat)
-import           ZkFold.Control.HApplicative (hliftA2, hpure)
-import           ZkFold.Data.ByteString      (Binary1)
-import           ZkFold.Data.HFunctor        (hmap)
-import           ZkFold.Data.Orphans         ()
-import           ZkFold.Data.Package         (pack)
-import           ZkFold.Data.Product         (fstP, sndP)
-import           ZkFold.Data.Vector          (Vector)
-import           ZkFold.Symbolic.Class       (Symbolic (WitnessField))
-
-type PayloadFunctor f = (Representable f, Binary (R.Rep f))
-
-type LayoutFunctor f = (Binary1 f, NFData1 f, PayloadFunctor f, Traversable f)
+import           Control.DeepSeq              (NFData, NFData1)
+import           Data.Binary                  (Binary)
+import           Data.Eq                      (Eq)
+import           Data.Function                (id, (.))
+import           Data.Functor                 (fmap)
+import           Data.Functor.Classes         (Eq1, Show1)
+import           Data.Functor.Rep             (Representable)
+import qualified Data.Functor.Rep             as R
+import           Data.Kind                    (Type)
+import           Data.Traversable             (Traversable)
+import           Data.Type.Equality           (type (~))
+import           GHC.Generics                 (V1, (:*:) (..), (:.:) (..))
+import qualified GHC.Generics                 as G
+import           GHC.Show                     (Show)
+
+import           ZkFold.Algebra.Number        (KnownNat)
+import           ZkFold.Control.HApplicative  (hpair)
+import           ZkFold.Data.ByteString       (Binary1)
+import           ZkFold.Data.HFunctor         (hmap)
+import           ZkFold.Data.HFunctor.Classes (HEq, HNFData, HShow)
+import           ZkFold.Data.Orphans          ()
+import           ZkFold.Data.Package          (pack, unpack)
+import           ZkFold.Data.Product          (fstP, sndP)
+import           ZkFold.Data.Vector           (Vector)
+import           ZkFold.Symbolic.Class        (Symbolic (..), embedW)
+
+-- | A newtype wrapper for deriving `Generic1` for symbolic data types.
+newtype Sym f c = Sym { runSym :: c f }
+
+deriving instance (HNFData c, NFData1 f) => NFData (Sym f c)
+deriving instance (HEq c, Eq1 f) => Eq (Sym f c)
+deriving instance (HShow c, Show1 f) => Show (Sym f c)
+
+instance G.Generic1 (Sym f) where
```

Comment:
In `Generic` deriving, there's a loopback instance for `G.Rec1`, so instead of
```haskell
instance G.Generic1 (Sym f)
instance G.Generic1 (Wit f)
instance LayoutFunctor f =>  GSymbolicData (Sym f)
instance PayloadFunctor f => GSymbolicData (Wit f)
```
it should be enough to have the following:
```haskell
instance SymbolicData x => GSymbolicData (G.Rec1 x)
instance LayoutFunctor f =>  SymbolicData (Sym f)
instance PayloadFunctor f => SymbolicData (Wit f)
```